syntax = "proto3";

//// https://protobuf.dev/reference/cpp/cpp-generated/#message
//option optimize_for = CODE_SIZE;

import "common.proto";
import "enums.proto";

package Odb.Lib.Protobuf;

//
//  EdaDataFile.h
//

message EdaDataFile {

    message FeatureIdRecord {
                
        enum Type {
            COPPER = 0;
            LAMINATE = 1;
            HOLE = 2;
        };
        
        optional Type type = 1;
        optional uint32 layerNumber = 2;
        optional uint32 featureNumber = 3;
    };   

    message NetRecord {

        message SubnetRecord {

            // common subnet enums
            enum Type {
                VIA = 0;
                TRACE = 1;
                PLANE = 2;
                TOEPRINT = 3;
            };

            // Plane subnet type enums
            enum FillType {
                SOLID = 0;
                OUTLINE = 1;
            };

            enum CutoutType {
                CIRCLE = 0;
                RECTANGLE = 1;
                OCTAGON = 2;
                EXACT = 3;
            };                    

            // common subnet fields
            optional Type type = 1;
            repeated FeatureIdRecord featureIdRecords = 2;

            // Plane subnet type fields
            optional FillType fillType = 3;
            optional CutoutType cutoutType = 4;
            optional double fillSize = 5;

            // Toeprint subnet type fields
            optional BoardSide side = 6;
            optional uint32 componentNumber = 7;
            optional uint32 toeprintNumber = 8;
            optional uint32 index = 9;
        };

        optional string name = 1;
        optional string attributesIdString = 2;
        optional uint32 index = 3;    

        repeated SubnetRecord subnetRecords = 4;
        repeated PropertyRecord propertyRecords = 5;
        map<string, string> attributeLookupTable = 6;  
    }

    message PackageRecord {    

        optional string name = 1;
        optional double pitch = 2;
        optional double xMin = 3;
        optional double yMin = 4;
        optional double xMax = 5;
        optional double yMax = 6;
        optional string attributesIdString = 7;

        repeated PinRecord pinRecords = 8;
        map<string, PinRecord> pinRecordsByName = 9;
        repeated PropertyRecord propertyRecords = 10;
        repeated OutlineRecord outlineRecords = 11;
        map<string, string> attributeLookupTable = 12;  

        message OutlineRecord {

            enum Type
            {
                Rectangle = 0;
                Circle = 1;
                Square = 2;
                Contour = 3;
            };

            optional Type type = 1;
            optional double lowerLeftX = 2;
            optional double lowerLeftY = 3;
            optional double width = 4;
            optional double height = 5;
            optional double xCenter = 6;
            optional double yCenter = 7;
            optional double halfSide = 8;
            optional double radius = 9;

            repeated ContourPolygon contourPolygons = 10;
        }

        message PinRecord {                      

            enum Type {
                ThroughHole = 0;
                Blind = 1;
                Surface = 2;
            };

            enum ElectricalType {
                Electrical = 0;
                NonElectrical = 1;
                Undefined = 2;
            };

            enum MountType {
                Smt = 0;
                RecommendedSmtPad = 1;
                MT_ThroughHole = 2;
                RecommendedThroughHole = 3;
                PressFit = 4;
                NonBoard = 5;
                Hole = 6;
                MT_Undefined = 7;  // default
            };

            optional string name = 1;
            optional Type type = 2;
            optional double xCenter = 3;
            optional double yCenter = 4;
            optional double finishedHoleSize = 5;    // unused, set to 0
            optional ElectricalType electricalType = 6;
            optional MountType mountType = 7;
            optional uint32 id = 8;
            optional uint32 index = 9;            
        }
    }   

    message FeatureGroupRecord  {
            
        optional string type = 1;

        repeated PropertyRecord propertyRecords = 2;        
        repeated FeatureIdRecord featureIdRecords = 3;
    }

    optional string path = 1;
    optional string units = 2;
    optional string source = 3;
    repeated string layerNames = 4;
    repeated string attributeNames = 5;
    repeated string attributeTextValues = 6;
    repeated NetRecord netRecords = 7;
    map<string, NetRecord> netRecordsByName = 8;
    repeated PackageRecord packageRecords = 9;
    map<string, PackageRecord> packageRecordsByName = 10;
    repeated PropertyRecord propertyRecords = 11;
    repeated FeatureGroupRecord featureGroupRecords = 12;
}